File: advent-of-code/day05/javascript/two.js
function countTotalFreshIds(inputText) {
  const lines = inputText.split(/\r?\n/);

  // ---------- 1) Parse only the fresh ranges (before blank line) ----------
  const freshRanges = [];

  for (const rawLine of lines) {
    const line = rawLine.trim();

    // Stop reading once we hit the blank line
    if (line === "") break;

    const [startStr, endStr] = line.split("-");
    const rangeStart = Number(startStr);
    const rangeEnd = Number(endStr);

    freshRanges.push([rangeStart, rangeEnd]);
  }

  if (freshRanges.length === 0) return 0;

  // ---------- 2) Sort ranges by start, then end ----------
  freshRanges.sort((a, b) => {
    if (a[0] !== b[0]) return a[0] - b[0];
    return a[1] - b[1];
  });

  // ---------- 3) Merge overlapping or touching ranges ----------
  const mergedRanges = [];

  let currentStart = freshRanges[0][0];
  let currentEnd = freshRanges[0][1];

  for (let i = 1; i < freshRanges.length; i++) {
    const [nextStart, nextEnd] = freshRanges[i];

    // Inclusive ranges:
    // If nextStart <= currentEnd + 1, they overlap or touch
    if (nextStart <= currentEnd + 1) {
      currentEnd = Math.max(currentEnd, nextEnd);
    } else {
      mergedRanges.push([currentStart, currentEnd]);
      currentStart = nextStart;
      currentEnd = nextEnd;
    }
  }

  // Push the final merged range
  mergedRanges.push([currentStart, currentEnd]);

  // ---------- 4) Sum up inclusive lengths ----------
  let totalFreshIds = 0;
  for (const [start, end] of mergedRanges) {
    totalFreshIds += end - start + 1;
  }

  return totalFreshIds;
}

// ---- AoC-style stdin usage ----
if (require.main === module) {
  const fs = require("fs");
  const input = fs.readFileSync(0, "utf8");
  console.log(countTotalFreshIds(input));
}

module.exports = { countTotalFreshIds };

File: advent-of-code/day05/python/two.py
from typing import List, Tuple

def count_total_fresh_ids(input_text: str) -> int:
    # 1) Parse only the first section (ranges), stop at blank line
    fresh_ranges: List[Tuple[int, int]] = []
    for raw_line in input_text.splitlines():
        line = raw_line.strip()
        if line == "":
            break
        start_str, end_str = line.split("-")
        fresh_ranges.append((int(start_str), int(end_str)))

    if not fresh_ranges:
        return 0

    # 2) Sort ranges by start, then end
    fresh_ranges.sort(key=lambda r: (r[0], r[1]))

    # 3) Merge overlapping/touching ranges
    merged_ranges: List[Tuple[int, int]] = []
    current_start, current_end = fresh_ranges[0]

    for next_start, next_end in fresh_ranges[1:]:
        if next_start <= current_end + 1:
            current_end = max(current_end, next_end)
        else:
            merged_ranges.append((current_start, current_end))
            current_start, current_end = next_start, next_end

    merged_ranges.append((current_start, current_end))

    # 4) Sum inclusive lengths
    total_fresh_ids = 0
    for start, end in merged_ranges:
        total_fresh_ids += end - start + 1

    return total_fresh_ids


if __name__ == "__main__":
    import sys
    print(count_total_fresh_ids(sys.stdin.read()))
